File: src/main/java/de/saschahlusiak/freebloks/client/JNIServer.java
package de.saschahlusiak.freebloks.client;

import android.util.Log;

import de.saschahlusiak.freebloks.model.Board;
import de.saschahlusiak.freebloks.model.GameMode;
import de.saschahlusiak.freebloks.model.Game;
import de.saschahlusiak.freebloks.model.Shape;

public class JNIServer {
	final static String tag = JNIServer.class.getSimpleName();

	/* unfortunately Runtime.availableProcessors() returns only the number of online cores */
	public static native int get_number_of_processors();

	private static native int native_run_server(int game_mode, int field_size_x, int field_size_y, int[] stones, int ki_mode, int ki_threads);

	private static native int native_resume_server(
			int field_size_x,
			int field_size_y,
			int current_player,
			int spieler[],
			int field_data[],
			int player_stone_data[],
			int game_mode,
			int ki_mode,
			int ki_threads);


	public static int runServer(Game game, GameMode gameMode, int field_size, int stones[], int ki_mode) {
		int ki_threads = get_number_of_processors();

		Log.d(tag, "spawning server with " + ki_threads + " threads");

		if (game == null)
			return native_run_server(gameMode.ordinal(), field_size, field_size, stones, ki_mode, ki_threads);
		else {
			final Board board = game.getBoard();
			int player_stones_available[] = new int[Shape.COUNT * 4];
			int i, j;

			for (i = 0; i < 4; i++)
				for (j = 0; j < Shape.COUNT; j++)
					player_stones_available[i * Shape.COUNT + j] = board.getPlayer(i).getStone(j).getAvailable();

			return native_resume_server(
					board.width,
					board.height,
					game.getCurrentPlayer(),
					game.getPlayerTypes(),
					board.getFields(),
					player_stones_available,
					gameMode.ordinal(),
					ki_mode,
					ki_threads);
		}
	}

	static {
		Log.d(tag, "loading server.so");
		System.loadLibrary("server");
	}
}

File: src/main/java/de/saschahlusiak/freebloks/client/JNIServer.kt
package de.saschahlusiak.freebloks.client

import android.util.Log
import de.saschahlusiak.freebloks.model.Game
import de.saschahlusiak.freebloks.model.GameMode
import de.saschahlusiak.freebloks.model.Shape

object JNIServer {
    private val tag = JNIServer::class.java.simpleName

    private external fun get_number_of_processors(): Int

    private external fun native_run_server(
        game_mode: Int,
        field_size_x: Int,
        field_size_y: Int,
        stones: IntArray?,
        ki_mode: Int,
        ki_threads: Int
    ): Int

    private external fun native_resume_server(
        field_size_x: Int,
        field_size_y: Int,
        current_player: Int,
        spieler: IntArray,
        field_data: IntArray,
        player_stone_data: IntArray,
        game_mode: Int,
        ki_mode: Int,
        ki_threads: Int
    ): Int

    @JvmStatic
    fun runServerForNewGame(gameMode: GameMode, size: Int, stones: IntArray?, kiMode: Int): Int {
        val threads = get_number_of_processors()
        Log.d(tag, "spawning server with $threads threads")
        return native_run_server(gameMode.ordinal, size, size, stones, kiMode, threads)
    }

    @JvmStatic
    fun runServerForExistingGame(game: Game, kiMode: Int): Int {
        val threads = get_number_of_processors()
        Log.d(tag, "spawning server with $threads threads")

        val board = game.board
        val playerStonesAvailable = IntArray(Shape.COUNT * 4)

        for (player in 0..3) {
            for (shape in 0 until Shape.COUNT) {
                playerStonesAvailable[player * Shape.COUNT + shape] = board.getPlayer(player).getStone(shape).available
            }
        }

        return native_resume_server(
            board.width,
            board.height,
            game.currentPlayer,
            game.playerTypes,
            board.fields,
            playerStonesAvailable,
            game.gameMode.ordinal,
            kiMode,
            threads
        )
    }

    init {
        Log.d(tag, "loading server.so")
        System.loadLibrary("server")
    }
}